#6232: Add embedding backwards op for training

TT-BrianLiu opened the pull request:
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/6232)

### Problem description
Embedding backwards is required for LLM training. This PR adds the new `embedding_bw` op on device, which uses the new `reshuffle_tile` LLK op introduced here: https://github.com/tenstorrent/tt-metal/issues/9817.

### What's changed
This PR adds the new `embedding_bw` op as a new `ttnn` op and removes the old implementation, which was faulty. 

### Checklist
- [x] Post commit CI passes
  - all post-commit: https://github.com/tenstorrent/tt-metal/actions/runs/10491100340
  - ttnn post-commit: https://github.com/tenstorrent/tt-metal/actions/runs/10491103136
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

razorback3 commented:
@ilkoo-lee

ilkoo-lee commented on ttnn/cpp/ttnn/operations/embedding_backward/device/kernels/compute/embedding_backward.cpp:
```diff
@@ -0,0 +1,83 @@
+        volatile uint32_t *chunk_addr_ptr;
+        cb_get_tile(cb_chunk_count_scratch, 0, &chunk_addr_ptr);
+
+        uint32_t chunk_count = chunk_addr_ptr[4]; // Need to shift because read ptr is off by 1 << 4 in BBE
```
@TT-BrianLiu Could you explain in more detail why the pointer was shifted by that amount? (BBE stands for budabackend, right?)

TT-BrianLiu replied:
@rdjogoTT Can you comment? It's somewhere there is a decrement and then `<< 4` in BBE so the read ptrs of the CB are misaligned between NCRISC and UNPACK.

TT-BrianLiu commented on ttnn/cpp/ttnn/operations/embedding_backward/device/kernels/compute/embedding_backward.cpp:
```diff
@@ -0,0 +1,83 @@
+    constexpr uint32_t max_tiles_per_core = get_compile_time_arg_val(0);
+    constexpr uint32_t input_height = get_compile_time_arg_val(1);
+
+		constexpr uint32_t cb_grad = tt::CB::c_in0;
```
Will fix in cleanup

TT-BrianLiu commented on ttnn/cpp/ttnn/operations/embedding_backward/device/embedding_backward_program_factory.cpp:
```diff
@@ -0,0 +1,181 @@
+        split_work_to_cores(grid_size, embedding_tiles);
+    uint32_t max_tiles_per_core = std::max(num_tiles_per_core_group_1, num_tiles_per_core_group_2);
+
+    log_info(LogType::LogOp, "Embedding hidden size tiles: {}", embedding_tiles);
```
Will do

TT-BrianLiu committed:
#6232: Add embedding backwards op for LLM training
- Remove old embedding_bw op from MCW
- Update embedding test to test arbitrary tokens within vocab size
- Update embedding test to test various input tensor dtypes
- Add TT-BrianLiu and yan-zaretskiy as codeowners of embedding_backward op
NOTE: This op only runs on Wormhole+ architecture since it uses a new llk
NOTE: ttnn.bfloat16 input indices only works for small vocab sizes
NOTE: output gradient (input to embedding_bw) dtype must be same as input gradient (output of embedding_bw)

tt-aho commented on ttnn/cpp/ttnn/operations/embedding_backward/embedding_backward_pybind.hpp:
```diff
@@ -0,0 +1,84 @@
+
+            Returns specific indices of the embedding table specified by the input tensor
+
+            /// TODO: Clean me up
+            :header: "Argument", "Description", "Data type", "Valid range", "Required"
+            "input", "Tensor containing rows we want", "UInt32 Tensor", "Each element greater than 0 and less than number of embeddings in table.  Shape [batch_size, 1, 1, num_rows]", "Yes"
+            "weights", "Entire embedding table", "Tensor", "Tensor shape is [1,1, num_embeddings, num_columns]. Num_columns must be divisible by 32.", "Yes"
+            "output_gradient", "Loss function gradient with respect to embedding output", "Tensor", "Tensor shape is [batch_size, 1, num_rows, num_columns]", "Yes"
+            "output_mem_config", "Layout of tensor in TT Accelerator device memory banks", "MemoryConfig", "Default is interleaved in DRAM", "No"
+            "output_dtype", "DataType of output tensor", "DataType", "Default is weights dtype", "No"
+            ///
```
Remove?

TT-BrianLiu replied:
I will add proper one

tt-aho commented on ttnn/cpp/ttnn/operations/embedding_backward/device/embedding_backward_program_factory.cpp:
```diff
@@ -0,0 +1,182 @@
+        auto index_dram_buffer = input_buffers.at(1);
+        auto output_dram_buffer = output_buffers.at(0);
+
+        for (const auto &core : cores) {
+            {
+                auto &runtime_args = GetRuntimeArgs(program, reader_kernel_id, core);
```
Change to this for better perf?
```suggestion
        auto &runtime_args_by_core = GetRuntimeArgs(program, reader_kernel_id);
        for (const auto &core : cores) {
            {
                auto &runtime_args = runtime_args_by_core[core.x][core.y];
```

tt-aho reviewed (approved):
Use 2024 in header for new files

dmakoviichuk-tt commented on ttnn/cpp/ttnn/operations/embedding_backward/device/embedding_backward_device_operation.cpp:
```diff
@@ -0,0 +1,88 @@
+        "Index tensor must be UINT32 or BFLOAT16");
+
+    TT_FATAL(
+        index_tensor_shape[1] == 1 && index_tensor_shape[2] == 1, "Only dim 0 && 3 for the index tensor can be non 1");
```
please add current shapes to the error messages

dmakoviichuk-tt commented on ttnn/cpp/ttnn/operations/embedding_backward/device/embedding_backward_device_operation.cpp:
```diff
@@ -0,0 +1,88 @@
+
+    TT_FATAL(
+        index_tensor_shape[-1] % TILE_WIDTH == 0,
+        "Number of columns in the index tensor must be divisible by tile width");
```
same idea. We need to include tile width abd shape

TT-BrianLiu replied:
Is there some convention i should follow?

ayerofieiev-tt replied:
@TT-BrianLiu , https://github.com/tenstorrent/tt-metal/blob/main/ErrorMessageBestPractices.md

dmakoviichuk-tt commented on ttnn/cpp/ttnn/operations/embedding_backward/embedding_backward_pybind.hpp:
```diff
@@ -0,0 +1,84 @@
+namespace ttnn::operations::embedding_backward {
+namespace py = pybind11;
+
+void py_module(py::module& module) {
```
please move pybind to cpp file

ayerofieiev-tt commented on ttnn/cpp/pybind11/operations/__init__.hpp:
```diff
@@ -91,6 +92,9 @@ void py_module(py::module& module) {
     auto m_embedding = module.def_submodule("embedding", "embedding operations");
     embedding::py_module(m_embedding);
 
+    auto m_embedding_backward = module.def_submodule("embedding_backward", "embedding backward operations");
```
why each op here is in own submodule? it does not make sense

TT-BrianLiu replied:
Embedding and embedding_bw are it's own ops right now. Should we move them somewhere else?

TT-BrianLiu replied:
Either way, we can revisit this later
